halo2_proofs/src/plonk/evaluation.rs

```diff
@@ -273,6 +276,39 @@ impl<C: CurveAffine> Evaluator<C> {
             ev.lookups.push(graph);
         }
 
+        // Shuffles
+        for shuffle in cs.shuffles.iter() {
+            let evaluate_lc = |expressions: &Vec<Expression<_>>, graph: &mut GraphEvaluator<C>| {
```

Comment:
Interesting approach here, I was wondering why we need to re-evaluate the `shuffle_expressions` here but then I realized if the expression is simple enough, it might be faster (and less memory) then converting the previously evaluated shuffled values into coset (requries `LagrangeCoeff -> Coeff -> ExtendedLagrangeCoeff `).

Perhaps in the future we have more concrete examples we could try to measure which one is faster, then decide which to use by the complexity of the expression in runtime.